README.md
# Elastic Search Archiver

This little command line tool is built to help you archive old logs in Elastic
Search, most likely logs Logstash put in there.

It will go and **flush indices transactions** then proceed to **close** those indices
so that they don't cost you CPU or memory, all this up to a certain duration,
the default is 2 weeks in the past. On top of that it has the ability to completly
delete (forever) indices up to a certain duration, default is not deletion.

## Usage

```
Usage of elastic-search-archiver:
  -archive=false: Pass flag to execute archival
  -archive-age=360h0m0s: Archive only indices old X time (Default is 15 days written 360h
  -delete=false: Pass flag to execute deletion
  -delete-age=720h0m0s: Delete only indices old X time (Default is 30 days written 720h
  -es-host="": Elastic Search host to delete indices from (required)
  -prefix="logstash-": Prefix behind ElasticSearch indices to delete or archive
```

## Developing

```bash
go get .
go run cmd/elastic-search-archiver/main.go
make build
```

## Liscence

(MIT)

Copyright (c) 2015 Frederic Gingras (frederic@gingras.cc)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## 0.1.1 [Upcomming]

### Added

### Changed

### Fixed

## 0.1.0 [2015-02-17]

### Added

- Basic feature set
  - Archival of indices up to a certain age
  - Deletion of indices up to a certain age
  - Specifing target ES host
  - Specifing prefix to scope deletion and archival



run.go
package esa

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"
	"time"
)

type ElasticSearchArchiver struct {
	c Config
}

func New(c Config) ElasticSearchArchiver {
	return ElasticSearchArchiver{c}
}

// Calls ElasticSearch with the prefix to figure out all collections concerned
// then will filter the one that will be deleted and the ones that will be
// archived then call the appropriate ElasticSearch endpoints to make that
// happen while logging what it does.
func (a ElasticSearchArchiver) Run() {
	err := a.c.Validate()
	if err != nil {
		log.Fatal(err)
	}

	baseUrl := a.c.ElasticSearchHost + "/"
	url := baseUrl + a.c.IndicePrefix + "*"

	// HTTP request to get all indices concerned
	res, err := http.Get(url)
	mustNotErr(err)
	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	mustNotErr(err)

	var data map[string]interface{}
	err = json.Unmarshal(body, &data)
	mustNotErr(err)

	toArchive := []string{}
	toDelete := []string{}
	archiveCutoff := time.Now().Add(-a.c.ArchiveAge)
	deleteCutoff := time.Now().Add(-a.c.DeleteAge)
	for key, _ := range data {
		keyWithoutPrefix := key[len(a.c.IndicePrefix):]
		keyDate, err := time.Parse("2006.01.02", keyWithoutPrefix)
		mustNotErr(err)

		if keyDate.Before(deleteCutoff) {
			toDelete = append(toDelete, key)
		} else if keyDate.Before(archiveCutoff) {
			toArchive = append(toArchive, key)
		}
	}

	log.Printf("target: %s\n", url)
	if a.c.Delete {
		for _, keyToDelete := range toDelete {
			req, err := http.NewRequest("DELETE", baseUrl+keyToDelete, nil)
			mustNotErr(err)
			_, err = http.DefaultClient.Do(req)
			mustNotErr(err)
			log.Println("deleted: " + keyToDelete[len(a.c.IndicePrefix):])
		}
	}
	if a.c.Archive {
		for _, keyToArchive := range toArchive {
			req, err := http.NewRequest("POST", baseUrl+keyToArchive+"/_flush", nil)
			mustNotErr(err)
			_, err = http.DefaultClient.Do(req)
			mustNotErr(err)
			req, err = http.NewRequest("POST", baseUrl+keyToArchive+"/_close", nil)
			mustNotErr(err)
			_, err = http.DefaultClient.Do(req)
			mustNotErr(err)
			log.Println("archived: " + keyToArchive[len(a.c.IndicePrefix):])
		}
	}
}

func mustNotErr(err error) {
	if err != nil {
		panic(err)
	}
}


main.go
package esa

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"
)

type ElasticSearchArchiver struct {
	c Config
}

func New(c Config) ElasticSearchArchiver {
	return ElasticSearchArchiver{c}
}

// Calls ElasticSearch with the prefix to figure out all collections concerned
// then will filter the one that will be deleted and the ones that will be
// archived then call the appropriate ElasticSearch endpoints to make that
// happen while logging what it does.
func (a ElasticSearchArchiver) Run() {
	err := a.c.Validate()
	if err != nil {
		log.Fatal(err)
	}

	url := a.c.ElasticSearchHost + "/" + a.c.IndicePrefix + "*"

	// HTTP request to get all indices concerned
	res, err := http.Get(url)
	mustNotErr(err)
	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	mustNotErr(err)

	var data map[string]interface{}
	err = json.Unmarshal(body, &data)
	mustNotErr(err)

	for key, _ := range data {
		log.Println(key)
	}
}

func mustNotErr(err error) {
	if err != nil {
		panic(err)
	}
}


config.go
package esa

import (
	"errors"
	"flag"
	"time"
)

type Config struct {
	Archive           bool
	ArchiveAge        time.Duration
	Delete            bool
	DeleteAge         time.Duration
	ElasticSearchHost string
	IndicePrefix      string
}

var (
	fArchive           *bool
	fArchiveAge        *time.Duration
	fDelete            *bool
	fDeleteAge         *time.Duration
	fElasticSearchHost *string
	fIndicePrefix      *string
)

func SetupConfigFlags() {
	fArchive = flag.Bool("archive", false, "Pass flag to execute archival")
	fArchiveAge = flag.Duration("archive-age", time.Hour*24*15, "Archive only indices old X time (Default is 15 days written 360h")
	fDelete = flag.Bool("delete", false, "Pass flag to execute deletion")
	fDeleteAge = flag.Duration("delete-age", time.Hour*24*30, "Delete only indices old X time (Default is 30 days written 720h")
	fElasticSearchHost = flag.String("es-host", "", "Elastic Search host to delete indices from (required)")
	fIndicePrefix = flag.String("prefix", "logstash-", "Prefix behind ElasticSearch indices to delete or archive")
}

func NewConfigFromFlags() Config {
	SetupConfigFlags()
	flag.Parse()
	return Config{
		*fArchive, *fArchiveAge, *fDelete, *fDeleteAge, *fElasticSearchHost, *fIndicePrefix,
	}
}

func (c Config) Validate() error {
	if !c.Archive && !c.Delete {
		return errors.New("Nothing to do, specify one or both of the archive and delete flags")
	}
	if c.Archive && c.ArchiveAge == 0 {
		return errors.New("Can't archive for indices old 0s, pass-in archive-age flag")
	}
	if c.Delete && c.DeleteAge == 0 {
		return errors.New("Can't archive for indices old 0s, pass-in archive-age flag")
	}
	if c.Archive && c.Delete && c.ArchiveAge >= c.DeleteAge {
		return errors.New("Age to archive from cannot be bigger than age to delete from")
	}
	if c.ElasticSearchHost == "" {
		return errors.New("ElasticSearch host flag is required")
	}
	return nil
}
